server/src/main/java/org/elasticsearch/transport/OutboundHandler.java

```diff
@@ -112,14 +104,22 @@ void sendErrorResponse(final Version nodeVersion, final TcpChannel channel, fina
     }
 
     private void sendMessage(TcpChannel channel, OutboundMessage networkMessage, ActionListener<Void> listener) throws IOException {
-        MessageSerializer serializer = new MessageSerializer(networkMessage, bigArrays);
-        SendContext sendContext = new SendContext(channel, serializer, listener, serializer);
-        internalSend(channel, sendContext);
+        final BytesStreamOutput bytesStreamOutput = new ReleasableBytesStreamOutput(bigArrays);
+        SendContext sendContext = new SendContext(channel, networkMessage, ActionListener.runBefore(listener, bytesStreamOutput::close));
+        final BytesReference message;
+        try {
+            message = networkMessage.serialize(bytesStreamOutput);
```

Comment:
Can we do this before we even create the `sendContext`? We haven't set the `startTime` by this point so the slow logging doesn't work, and logging `send message failed` is a bit weird on a serialization exception since it's probably a version incompatibility that caused the exception.

Reply:
Good point :) saves us the whole dance with the `-1` message size and allows for logging a more reasonable error as well here https://github.com/elastic/elasticsearch/pull/72442/commits/85b51424e3b1c2c19da947613b818aae39763276

... I wonder if the exception path here should just `assert false` actually. I can't think of a non-bug reason this would ever throw.

Reply:
We legitimately throw exceptions during serialisation if the target node is running an older version which cannot understand the request, typically because it uses features that are only supported in newer versions. Failing requests like that in a mixed-version cluster is ok, stuff like this:

https://github.com/elastic/elasticsearch/blob/a92a647b9f17d1bddf5c707490a19482c273eda3/server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/GetSnapshotsRequest.java#L82-L83

Reply:
Just spotted that the message is now logged at `error` here, that's overly dramatic IMO let's stick to `warn`.

Reply:
Right ... I think this is the third time I forget about that spot and you link me to it now :)

-> Moved this back to `warn` :)